proposals/0000-complete-set-signatures.rst

```diff
@@ -0,0 +1,420 @@
+``safeHead``.
+But in doing so, the following ``unsafeHead`` will not emit a pattern match
+warning anymore:
+
+::
+
+ unsafeHead :: [a] -> a
+ unsafeHead (x :< _) = x
+
+Urgh! We somehow want to say that the singleton COMPLETE set only applies
+to type constructors satisfying ``IsInfinite``. But the type signature
+syntax doesn't allow us to constrain on ``IsInfinite``! The only way out
+is to declare the COMPLETE signature for all concrete type constructors
+such as ``Stream``:
+
+::
+
+ {-# COMPLETE (:<) :: Stream #-}
+
+And here goes repeating that declaration for all type constructors again, for us
+as well as the users of our library. I'd much rather write
+
+::
+
+ {-# COMPLETE (:<), Empty :: IsList l => l a #-}
+ {-# COMPLETE (:<) :: IsInfinite l => l a #-}
+
+Once, inside the library. The constraint on the
+latter makes sure it doesn't apply to the match
+type in ``unsafeHead``, while it applies to any
+match type that satisfies ``IsInfinite`` that a
+user might write.
+
+That is the new feature that I propose. Note the
+specification of the full result type instead of
+just the name of a data type constructor: It's a
+breaking change with the current syntax, but one
+that is trivially flagged by a kinding error.
+
+Proposed Change Specification
+-----------------------------
+
+Syntax
+======
+
+Extend Syntax so that we are able to write the example from the previous section
+
+::
+
+ {-# COMPLETE (:<), Empty :: IsList l => l a #-} -- Context is allowed
+ {-# COMPLETE (:<) :: Stream a #-}               -- TyCon App is allowed
+ {-# COMPLETE (:<), Empty :: String #-}          -- Even type synonyms
+
+Since there is no formal grammar for COMPLETE pragmas, here's how I propose to
+change the happy grammar from
+
+::
+
+ sigdecl
+   : ...
+   | '{-# COMPLETE' con_list opt_tyconsig  '#-}'
+ opt_tyconsig
+   : {- -}
+   | '::' gtycon
+ gtycon
+   : ntgtycon
+   | '(' ')'
+   | '(#' '#)'
+
+to
+
+::
+
+ sigdecl
+   : ...
+   | '{-# COMPLETE' con_list opt_restysig  '#-}'
+ opt_restysig
+   : {- -}
+   | '::' ctype
+ atype
+   : ntgtycon
+   | '(' ')'
+   | '(#' '#)'
+   | ... many more ...
+
+Where a ``ctype`` is a "for-all type", according to GHC's happy parser. It's the
+same grammatical sort that constitutes the RHS of a ``::`` in a type signature
+and since it constrains the result type of the relevant ConLikes, I'd like to
+call it a "result type signature".
+
+Note that this change accepts strictly more syntax, because
+
+  - ``atype`` accepts a superset of ``gtycon``
+  - ``atype`` can be derived from ``ctype`` (via ``type``, ``btype``, ``infixtype``, ``ftype``)
+
+Typing
+======
+
+The ``ctype`` in the result type signature is renamed and kind-checked like a
+``ctype`` occuring in an ordinary type signature.
+
+Examples for well-typed ``ctype``s:
+
+::
+
+ [a]
+ IsInfinite l => l a
+ String
+ (a ~ Int, Semigroup a) => a
+ MPTC a b => a
+
+Examples for invalid ``ctype``s:
```

Comment:
When you say `ctype`, I’m thinking about the non-terminal defined in the grammar. But syntactically, `Stream`, `(->)`, and `"Symbol"` are all valid `ctype`s.

I think you meant to say “Examples of ill-kinded result type signatures”